### PR TITLE
feat: Fire ExpandEvent in expandRecursively

### PR DESCRIPTION
**Pull request by TatuLund**
fixes: #2461

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridExpandRecursivelyPage.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        grid.addComponentHierarchyColumn(value -> new Label(value))
+                .setHeader("Name").setAutoWidth(true).setFlexGrow(0);
+        grid.addComponentColumn(value -> new Label(new Date().toString()))
+                .setHeader("Datum");
+        grid.addComponentColumn(value -> new Label(new Date().toString()))
+                .setHeader("Datum 2");
+
+        grid.addExpandListener(event -> {
+            Label label = new Label("Expanded");
```
Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridExpandRecursivelyPage.java:46]: Please don't use `Label` element for text. See the description for V10+ Label at https://vaadin.com/api/platform/10.0.19/com/vaadin/flow/component/html/Label.html

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java [hunk 2]
```diff
@@ -796,6 +796,10 @@ public void expandRecursively(Stream<T> items, int depth) {
     public void expandRecursively(Collection<T> items, int depth) {
         getDataCommunicator()
                 .expand(getItemsWithChildrenRecursively(items, depth));
+        // Call with client roundtrip as expansion takes a while.
+        // This ensures that event is fired after exapnsion is complete.
+        getElement().executeJs("return 1;").then(result -> fireEvent(
+                new ExpandEvent<T, TreeGrid<T>>(this, false, items)));
```
Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java:802]: The `items` here only contains the items at the root of the recursive expansion. Sub-items that may also get expanded aren't included in the event.

Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java:802]: The JavaDoc for this method says `This method will <i>not</i> fire events for expanded nodes.`

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java:802]: I will use null instead

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java:802]: I changed the issue type to enhancement

[the remark this answers is not in the document]
Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java:802]: I will remove the delay

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridExpandRecursivelyPage.java (hunk 1 above)
Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridExpandRecursivelyPage.java:46]: Will change to Span

**Comment by yuriy-fix** [comment]
Thank you for your contribution!
The original issue was fixed with this fix: https://github.com/vaadin/flow-components/pull/2540. Closing as enhancements is not needed anymore.

**Comment by TatuLund** [comment]
Eventually there could be other use cases, which would make this feature useful.

**Comment by tomivirkki** [comment]
> Eventually there could be other use cases, which would make this feature useful.

The `ExpandEvent` type doesn't fit well for the case of recursively expanding items. The suggested convention to use `null` for the `items` property doesn't sound right, because then the event doesn't really tell you anything else than "something in the grid was expanded recursively". If an event is needed for recursive expansion, there should probably be a dedicated event type for it with perhaps `rootItems` property. As such, the PR is also breaking, because methods that explicitly documented not to fire `ExpandEvent` events, would start firing them.

**Pull request closed without merging**